commands/management/add_external.go

```diff
@@ -0,0 +1,132 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package management
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/percona/pmm/api/managementpb/json/client"
+	"github.com/percona/pmm/api/managementpb/json/client/external"
+
+	"github.com/percona/pmm-admin/agentlocal"
+	"github.com/percona/pmm-admin/commands"
+)
+
+var addExternalResultT = commands.ParseTemplate(`
+External Service added.
+Service ID  : {{ .Service.ServiceID }}
+Service name: {{ .Service.ServiceName }}
+`)
+
+type addExternalResult struct {
+	Service *external.AddExternalOKBodyService `json:"service"`
+}
+
+func (res *addExternalResult) Result() {}
+
+func (res *addExternalResult) String() string {
+	return commands.RenderTemplate(addExternalResultT, res)
+}
+
+type addExternalCommand struct {
+	RunsOnNodeID   string
+	ServiceName    string
+	Username       string
+	Password       string
+	Scheme         string
+	MetricPath     string
+	ListenPort     uint16
+	NodeID         string
+	Environment    string
+	Cluster        string
+	ReplicationSet string
+	CustomLabels   string
+}
+
+func (cmd *addExternalCommand) Run() (commands.Result, error) {
+	customLabels, err := commands.ParseCustomLabels(cmd.CustomLabels)
+	if err != nil {
+		return nil, err
+	}
+
+	if cmd.RunsOnNodeID == "" || cmd.NodeID == "" {
+		status, err := agentlocal.GetStatus(agentlocal.DoNotRequestNetworkInfo)
+		if err != nil {
+			return nil, err
+		}
+		if cmd.RunsOnNodeID == "" {
+			cmd.RunsOnNodeID = status.NodeID
+		}
+		if cmd.NodeID == "" {
+			cmd.NodeID = status.NodeID
+		}
+	}
+
+	params := &external.AddExternalParams{
+		Body: external.AddExternalBody{
+			RunsOnNodeID:   cmd.RunsOnNodeID,
+			ServiceName:    cmd.ServiceName,
+			Username:       cmd.Username,
+			Password:       cmd.Password,
+			Scheme:         cmd.Scheme,
+			MetricPath:     cmd.MetricPath,
+			ListenPort:     int64(cmd.ListenPort),
+			NodeID:         cmd.NodeID,
+			Environment:    cmd.Environment,
+			Cluster:        cmd.Cluster,
+			ReplicationSet: cmd.ReplicationSet,
+			CustomLabels:   customLabels,
+		},
+		Context: commands.Ctx,
+	}
+	resp, err := client.Default.External.AddExternal(params)
+	if err != nil {
+		return nil, err
+	}
+
+	return &addExternalResult{
+		Service: resp.Payload.Service,
+	}, nil
+}
+
+// register command
+var (
+	AddExternal  = new(addExternalCommand)
+	AddExternalC = AddC.Command("external", "Add External to monitoring")
+)
+
+func init() {
+	hostname, _ := os.Hostname()
+	serviceName := hostname + "-external"
+	serviceNameHelp := fmt.Sprintf("Service name (autodetected default: %s)", serviceName)
+	AddExternalC.Flag("service-name", serviceNameHelp).Default(serviceName).StringVar(&AddExternal.ServiceName)
```

Comment:
Looks like this flag conflicts with the same in add.go

Reply:
good point